accepted/future-releases/0546-patterns/exhaustiveness.md

```diff
@@ -246,772 +251,854 @@ also be an instance of one of the known set of subtypes. That gives us the
+  Pip _ => 'pip',
+  Face _ => 'face'
+};
 ```
 
-And now it's easier to see that `Jack` subtracts the first arm leaving:
+It's hard to see how we might easily tell that this switch is exhaustive. After
+the algorithm has looked at the first `Pip _` case, what does it know about the
+set of values that have been covered? More to the point, how does it know what's
+*left*? If we supported set operations on spaces, we could say that the
+remaining values are the `Card` space minus the `Pip` space. But we don't want
+to have to define that.
 
-```
-Queen(suit: heart)|King(suit: heart)
+However, we can observe that because `Card` is sealed, we know that any instance
+of `Card` will be an instance of one of its direct sealed subtypes. That means
+that if this pair of switch statements is exhaustive, then the above switch must
+be too:
+
+```dart
+test(Pip pip) => switch (pip) {
+  Pip _ => 'pip',
+  Face _ => 'face'
+};
+
+test(Face face) => switch (face) {
+  Pip _ => 'pip',
+  Face _ => 'face'
+};
 ```
 
-If the left type is nullable and the right type is `Null` or non-nullable, then
-expanding expands the nullable type to `Null` and the underlying type, as if the
-nullable type was a sealed supertype of the underlying type and `Null`:
+Note that each switch is only matching one of the sealed subtypes now. It's now
+trivial to see that the first switch is exhaustive because every matched value
+will be an instance of `Pip` and the first case covers every single instance of
+`Pip`. (The second case is unreachable and pointless, but harmless.) Likewise,
+the second switch is clearly exhaustive because every value is a `Face` and the
+second case matches all instances of `Face`.
 
-```
-Face? - Face  expands to:  Face|Null - Face = Null
-Jack? - Null  expands to:  Jack|Null - Null = Jack
-```
+In the algorithm, when it looks at a space, it first *expands* it. If the type
+of a space is a sealed type, enum type, or `bool` (which is basically an enum
+type), then it replaces the space with a set of more precise spaces that cover
+the original space. Then it processes those each independently. If the cases are
+all exhaustive over every one of the expanded spaces, then they are exhaustive
+over the original unexpanded space too.
 
-Likewise, if the left type is `FutureOr<T>` for some type `T` and the right type
-is a subtype of `Future` or `T`, then expanding expands the `FutureOr<T>` to
-`Future<T>|T`.
+To expand a space `s` with type `T`, we create a new set of spaces that share
+the same properties as `s` but with possibly different types and restrictions.
+The resulting spaces are:
 
-```
-FutureOr<int> - int     expands to:  Future<int>|int - int = Future<int>
-FutureOr<int> - Future  expands to:  Future<int>|int - Future = int
-```
+*   **The declaration `D` of `T` is a `sealed` type:**
 
-### Subtraction
+    1.  For each declaration `C` in the library of `D` that has `D` in an
+        `implements`, `extends`, or `with` clause:
```

Comment:
Here I want to restrict it to:

... declaration `C` in the library of `D` that has `D` as a direct superdeclaration through an `implements`, `extends` or `with` clause _(I agree with excluding `on` types too)_
and which does not have any other superdeclaration *P* which has *D* as a proper superdeclaration.

This avoids weird behavior for unnecessary implements clauses:
```dart
sealed class S {}
class C extend S {}
class D extends S {}
class E extends D implements S {}
```
Here class `E` should *not* be included, because it's a proper subtype of a proper subtype of `S` (if `D` was not `sealed`, it would be included transitiviely in exhausting `S`, but that's not important here).

It's not necessary to include `E` in the set of spaces for `S` here (exhaustiveness of `S` requires exhausting `D` anyway), so it just complicates everything.

(That is: The *longest* path from the subclass to the sealed class in the class declaration hierarchy has length 1. Internal anonymous mixin applications don't count since we use "declaration" here.
Or as you say below, "a direct subtype" - ignoring anonymous mixin types as usual.)

Reply:
I don't have a strong opinion about this, but I think it's harmless to include redundant direct subtypes and I think the proposal's a little simpler and clearer to not make an exception for them.